test/Microsoft.AspNetCore.Razor.Test.Common/Language/Legacy/SyntaxTreeNodeSerializer.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy
+{
+    public static class SyntaxTreeNodeSerializer
+    {
+        internal static string Serialize(SyntaxTreeNode node)
+        {
+            using (var writer = new StringWriter())
+            {
+                var walker = new Walker(writer);
+                walker.Visit(node);
+
+                return writer.ToString();
+            }
+        }
+
+        private class Walker : SyntaxTreeNodeWalker
+        {
+            private readonly SyntaxTreeNodeWriter _visitor;
+            private readonly TextWriter _writer;
+
+            public Walker(TextWriter writer)
+            {
+                _visitor = new SyntaxTreeNodeWriter(writer);
+                _writer = writer;
+            }
+
+            public TextWriter Writer { get; }
+
+            public override void Visit(SyntaxTreeNode node)
+            {
+                _visitor.Visit(node);
+                _writer.WriteLine();
+
+                if (node is Block block)
```

Comment:
Any reason this piece isn't just part of the `SyntaxTreeNodeWriter`? If you had that there wouldn't be any need for this walker class.

Reply:
Because this and `SyntaxTreeNodeVerifier` both use the `SyntaxTreeNodeWriter` and these two have slightly different implementations of Visit.